File: backend/app/schemas/workout.py
from pydantic import BaseModel, Field
from typing import List, Optional, Any
from uuid import UUID
from datetime import datetime
from app.models.workout import EquipmentEnum, DifficultyEnum, SessionStatusEnum


# --- Exercise Schemas ---
class ExerciseBase(BaseModel):
    name: str
    category: str
    primary_muscle: str
    secondary_muscles: List[str] = Field(default_factory=list)
    equipment: EquipmentEnum
    difficulty: DifficultyEnum
    instructions: List[str] = Field(default_factory=list)
    gif_url: Optional[str] = None
    calories_per_minute_est: float = 5.0


class ExerciseResponse(ExerciseBase):
    id: UUID

    model_config = {"from_attributes": True}


# --- Set Detail in JSONB ---
class SetDetail(BaseModel):
    set_number: int = Field(..., ge=1)
    reps: int = Field(..., ge=0)
    weight_kg: float = Field(default=0.0, ge=0)
    completed: bool = True


# --- Workout Exercise Log Schemas ---
class WorkoutExerciseLogCreate(BaseModel):
    exercise_id: UUID
    order_index: int = Field(default=1, ge=1)
    sets_data: List[SetDetail] = Field(default_factory=list)
    target_rest_seconds: int = Field(default=90, ge=0)


class WorkoutExerciseLogResponse(BaseModel):
    id: UUID
    workout_log_id: UUID
    exercise_id: UUID
    order_index: int
    sets_data: List[Any] = Field(default_factory=list)
    target_rest_seconds: int
    exercise: Optional[ExerciseResponse] = None

    model_config = {"from_attributes": True}


# --- Workout Log Schemas ---
class WorkoutLogCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    routine_tag: Optional[str] = Field(default=None, max_length=100)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    total_volume_kg: Optional[float] = Field(default=0.0, ge=0)
    calories_burned: Optional[float] = Field(default=0.0, ge=0)
    notes: Optional[str] = None
    status: SessionStatusEnum = SessionStatusEnum.in_progress
    exercise_logs: List[WorkoutExerciseLogCreate] = Field(default_factory=list)


class WorkoutLogUpdate(BaseModel):
    title: Optional[str] = None
    routine_tag: Optional[str] = None
    completed_at: Optional[datetime] = None
    total_volume_kg: Optional[float] = None
    calories_burned: Optional[float] = None
    notes: Optional[str] = None
    status: Optional[SessionStatusEnum] = None
    exercise_logs: Optional[List[WorkoutExerciseLogCreate]] = None


class WorkoutLogResponse(BaseModel):
    id: UUID
    user_id: UUID
    title: str
    routine_tag: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    total_volume_kg: float = 0.0
    calories_burned: float = 0.0
    notes: Optional[str] = None
    status: SessionStatusEnum
    exercise_logs: List[WorkoutExerciseLogResponse] = Field(default_factory=list)

    model_config = {"from_attributes": True}

File: frontend/src/services/workout.ts
import api from './api';

// ═══════════════════════════════════════════════════════════════
// TypeScript Interfaces — mirroring Pydantic schemas
// ═══════════════════════════════════════════════════════════════

export interface Exercise {
  id: string;
  name: string;
  category: string;
  primary_muscle: string;
  secondary_muscles: string[];
  equipment: string;
  difficulty: string;
  instructions: string[];
  gif_url: string | null;
  calories_per_minute_est: number;
}

export interface SetDetail {
  set_number: number;
  reps: number;
  weight_kg: number;
  completed: boolean;
}

export interface WorkoutExerciseLogCreate {
  exercise_id: string;
  order_index: number;
  sets_data: SetDetail[];
  target_rest_seconds: number;
}

export interface WorkoutExerciseLogResponse {
  id: string;
  workout_log_id: string;
  exercise_id: string;
  order_index: number;
  sets_data: SetDetail[];
  target_rest_seconds: number;
  exercise: Exercise | null;
}

export interface WorkoutLogCreate {
  title: string;
  routine_tag?: string;
  started_at?: string;
  completed_at?: string;
  total_volume_kg?: number;
  calories_burned?: number;
  notes?: string;
  status: 'in_progress' | 'completed' | 'skipped';
  exercise_logs: WorkoutExerciseLogCreate[];
}

export interface WorkoutLogResponse {
  id: string;
  user_id: string;
  title: string;
  routine_tag: string | null;
  started_at: string | null;
  completed_at: string | null;
  total_volume_kg: number;
  calories_burned: number;
  notes: string | null;
  status: 'in_progress' | 'completed' | 'skipped';
  exercise_logs: WorkoutExerciseLogResponse[];
}

// ═══════════════════════════════════════════════════════════════
// Exercise Catalog Filters
// ═══════════════════════════════════════════════════════════════

export interface ExerciseFilters {
  primary_muscle?: string;
  equipment?: string;
  category?: string;
  difficulty?: string;
  q?: string;
}

// ═══════════════════════════════════════════════════════════════
// API Functions
// ═══════════════════════════════════════════════════════════════

export async function fetchExercises(filters?: ExerciseFilters): Promise<Exercise[]> {
  const params = new URLSearchParams();
  if (filters?.primary_muscle) params.append('primary_muscle', filters.primary_muscle);
  if (filters?.equipment) params.append('equipment', filters.equipment);
  if (filters?.category) params.append('category', filters.category);
  if (filters?.difficulty) params.append('difficulty', filters.difficulty);
  if (filters?.q) params.append('q', filters.q);

  const res = await api.get('/exercises', { params });
  return res.data;
}

export async function fetchExerciseById(id: string): Promise<Exercise> {
  const res = await api.get(`/exercises/${id}`);
  return res.data;
}

export async function submitWorkout(payload: WorkoutLogCreate): Promise<WorkoutLogResponse> {
  const res = await api.post('/workouts', payload);
  return res.data;
}

export async function fetchWorkoutHistory(): Promise<WorkoutLogResponse[]> {
  const res = await api.get('/workouts');
  return res.data;
}

export async function fetchWorkoutById(id: string): Promise<WorkoutLogResponse> {
  const res = await api.get(`/workouts/${id}`);
  return res.data;
}

export async function deleteWorkout(id: string): Promise<void> {
  await api.delete(`/workouts/${id}`);
}
